src/collapse/vsprof.rs

```diff
@@ -0,0 +1,222 @@
+                    // the previous node (in this case `B()`) from the current node (in this case
+                    // `A()`. This leaves `A()` with the correct number of 500 samples.
+                    //
+                    // If the previous number of calls is equal to the current number of calls, we
+                    // don't want to write the current top node, because that would duplicate the
+                    // number of samples for the current node.
+                    let mut prev_number_of_calls = 0;
+                    for _ in 0..(prev_depth - depth + 1) {
+                        if prev_number_of_calls != self.stack.last().unwrap().1 {
+                            self.write_stack(occurences);
+                        }
+                        prev_number_of_calls = self.stack.pop().unwrap().1;
+
+                        if self.stack.is_empty() {
+                            break;
+                        }
+
+                        let last = self.stack.len() - 1;
+                        let number_of_calls = &self.stack[last].1;
+                        if prev_number_of_calls < *number_of_calls {
+                            self.stack[last].1 -= prev_number_of_calls;
+                        }
+                    }
+
+                    self.stack
+                        .push((function_name.to_string(), number_of_calls));
+                }
+            }
+        } else {
+            return invalid_data_error!("Unable to parse function name from line:\n{}", line);
+        }
+
+        Ok(())
+    }
+
+    // Store the current stack in `occurences`
+    fn write_stack(&self, occurrences: &mut Occurrences) {
+        if let Some(nsamples) = self.stack.last().map(|(_, n)| *n) {
+            if nsamples > 0 {
+                let functions: Vec<_> = self.stack.iter().cloned().map(|(f, _)| f).collect();
+                occurrences.insert(functions.join(";"), nsamples);
+            }
+        }
+    }
+}
+
+/// Gets the number from the start of the line. This can either be a number <1000, in which case the
+/// line doesn't contain double quotes, or the number can be >1000, in which case the line does
+/// contain double quotes. In both cases `line` may start with a leading comma, which will be
+/// ignored.
+fn get_next_number(line: &str) -> io::Result<(usize, &str)> {
+    // Trim the leading comma, if any
+    let line = line.strip_prefix(',').unwrap_or(line);
+
+    let mut remove_leading_comma = false;
+    let split = if let Some(line) = line.strip_prefix('"') {
+        remove_leading_comma = true;
+        line.split_once('"')
+    } else {
+        line.split_once(',')
+    };
+
+    if let Some((num, remainder)) = split {
+        // Remove any thousands separators which may be present
+        // TODO: If Visual Studio were to output numbers with a `.` as thousands separators, this
+        //       would fail
+        let num = num.replace(',', "");
```

Comment:
This will also allocate a string, which seems sad. Here's a counter-proposal:
```rust
fn parse(field: &str) -> std::io::Result<usize> {
    let thousands = field.split(|c: char| !c.is_ascii_digit());
    let mut n = 0;
    for thousand in thousands {
        // return invalid data error here instead if you prefer
        if n == 0 {
            assert!(thousand.len() <= 3);
        } else {
            assert_eq!(thousand.len(), 3);
        }
        n *= 1000;
        n += thousand.parse::<usize>()?;
    }
    Ok(n)
}
```